src/psyclone/psyir/transformations/acc_update_trans.py

```diff
@@ -0,0 +1,403 @@
+
+        # TODO #1872: as a workaround for the lack of precise array access
+        # descriptions, we currently overapproximate dependencies by adding any
+        # variables being written to the input set (in addition to the output
+        # set). The aim is to avoid situations where the host only writes an
+        # array slice and the respective update device directive would then
+        # incorrectly overwrite the entire array on the device, including the
+        # data outside the slice. If or when this is no longer needed, the step
+        # of the algorithm below dealing with if stmts will need updating.
+        inputs.update(outputs)
+
+        # TODO #1732: for CodeBlock nodes, until we have semantic information
+        # on the statements involved, we conservatively assume every referenced
+        # variable is both an input and an output.
+        # TODO #1872: CodeBlock.get_symbol_names might include function names
+        # mistaken for array names. However, as these do not correspond to any
+        # arrays, the compiler should be able to ignore them in the directive.
+        for node in node_list:
+            if isinstance(node, CodeBlock):
+                for symbol_name in node.get_symbol_names():
+                    inputs.add(Signature(symbol_name))
+                    outputs.add(Signature(symbol_name))
+
+        # Copy any data that is read by this region to the host (resp. device)
+        # if it is on the device (resp. host).
+        IN, OUT = 0, 1
+        for idx, host_sig in enumerate((inputs, outputs)):
+            if idx == IN:     # inputs
+                node_index = 0
+                node_offset = 0
+                direction = "host"
+                tentative_update_pos = "beg"
+            elif idx == OUT:  # outputs
+                node_index = -1
+                node_offset = 1
+                direction = "device"
+                tentative_update_pos = "end"
+
+            child = node_list[node_index]
+            sched = child.parent
+            # Since the supplied nodes may be the children of an IfBlock or a
+            # Loop, we get the deepest ancestor that is a child of a Schedule.
+            while not isinstance(sched, Schedule):
+                child, sched = child.parent, sched.parent
+
+            # While there are variables from this host region which have not
+            # been placed in an update directive, we check if they need to be
+            # placed at the current schedule, in which case they are, or
+            # if we can bring them up to the next deepest ancestor schedule.
+            # The main goal is to avoid placing update directives inside loop
+            # statements, but the algorithm also tries to minimise the total
+            # number of update directives by promoting the coalescence of
+            # neighbouring directives.
+            while host_sig:
+                # Most conservative position for a new update directive at this
```

Comment:
It could be that the body of this loop might also be best replaced with a method - it aids readability and testing.